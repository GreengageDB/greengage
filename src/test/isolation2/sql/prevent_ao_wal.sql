-- For AO/AOCO tables, their WAL records are only
-- generated for replication purposes (they are not used for crash
-- recovery because AO/AOCO table operations are crash-safe). To decrease
-- disk space usage and to improve performance of AO/AOCO operations, we
-- suppress generation of XLOG_APPENDONLY_INSERT and
-- XLOG_APPENDONLY_TRUNCATE WAL records when wal_level=minimal is
-- specified.
-- This test is supposed to ensure that XLOG_APPENDONLY_INSERT and
-- XLOG_APPENDONLY_TRUNCATE WAL records are not generated when
-- wal_level=minimal is set.
-- Because on mirrored cluster primary segments have replication slots
-- and that conflict with the wal_level=minimal GUC
-- we connect to coordinator in utility mode for AO/AOCO operations and
-- validate WAL records on the coordinator.

-- start_matchignore
-- m/pg_xlogdump: FATAL:  error in WAL record at */
-- m/.*The 'DISTRIBUTED BY' clause determines the distribution of data*/
-- m/.*Table doesn't have 'DISTRIBUTED BY' clause*/
-- end_matchignore

-- Create tables (AO, AOCO)
-1U: CREATE TABLE ao_foo (n int) WITH (appendonly=true);
-1U: CREATE TABLE aoco_foo (n int, m int) WITH (appendonly=true, orientation=column);

-- Switch WAL file
-1U: SELECT true FROM pg_switch_xlog();
-- Insert data (AO)
-1U: INSERT INTO ao_foo SELECT generate_series(1,10);
-- Insert data (AOCO)
-1U: INSERT INTO aoco_foo SELECT generate_series(1,10), generate_series(1,10);
-- Delete data and run vacuum (AO)
-1U: DELETE FROM ao_foo WHERE n > 5;
-1U: VACUUM ao_foo;
-- Delete data and run vacuum (AOCO)
-1U: DELETE FROM aoco_foo WHERE n > 5;
-1U: VACUUM aoco_foo;
-1Uq:

-- Validate wal records (mirrorless setting has alternative answer file for this since wal_level is already minimal)
! last_wal_file=$(psql -At -c "SELECT pg_xlogfile_name(pg_current_xlog_location())" postgres) && pg_xlogdump ${last_wal_file} -p ${MASTER_DATA_DIRECTORY}/pg_xlog -r appendonly;

-- *********** Set wal_level=minimal **************
!\retcode gpconfig -c wal_level -v minimal --masteronly --skipvalidation;
-- Set max_wal_senders to 0 because a non-zero value requires wal_level >= 'archive'
!\retcode gpconfig -c max_wal_senders -v 0 --masteronly --skipvalidation;
-- Restart QD
!\retcode pg_ctl -l /dev/null -D $MASTER_DATA_DIRECTORY restart -w -t 600 -m fast;

-- Switch WAL file
-1U: SELECT true FROM pg_switch_xlog();
-- Insert data (AO)
-1U: INSERT INTO ao_foo SELECT generate_series(1,10);
-- Insert data (AOCO)
-1U: INSERT INTO aoco_foo SELECT generate_series(1,10), generate_series(1,10);
-- Delete data and run vacuum (AO)
-1U: DELETE FROM ao_foo WHERE n > 5;
-1U: VACUUM ao_foo;
-- Delete data and run vacuum (AOCO)
-1U: DELETE FROM aoco_foo WHERE n > 5;
-1U: VACUUM aoco_foo;

-- Validate wal records
! last_wal_file=$(psql -At -c "SELECT pg_xlogfile_name(pg_current_xlog_location())" postgres) && pg_xlogdump ${last_wal_file} -p ${MASTER_DATA_DIRECTORY}/pg_xlog -r appendonly;

-1U: DROP TABLE ao_foo;
-1U: DROP TABLE aoco_foo;

-- Reset wal_level
!\retcode gpconfig -r wal_level --masteronly --skipvalidation;
-- Reset max_wal_senders
!\retcode gpconfig -r max_wal_senders --masteronly --skipvalidation;
-- Restart QD
!\retcode pg_ctl -l /dev/null -D $MASTER_DATA_DIRECTORY restart -w -t 600 -m fast;
