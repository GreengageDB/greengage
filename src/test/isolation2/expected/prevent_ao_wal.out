-- For AO/AOCO tables, their WAL records are only
-- generated for replication purposes (they are not used for crash
-- recovery because AO/AOCO table operations are crash-safe). To decrease
-- disk space usage and to improve performance of AO/AOCO operations, we
-- suppress generation of XLOG_APPENDONLY_INSERT and
-- XLOG_APPENDONLY_TRUNCATE WAL records when wal_level=minimal is
-- specified.
-- This test is supposed to ensure that XLOG_APPENDONLY_INSERT and
-- XLOG_APPENDONLY_TRUNCATE WAL records are not generated when
-- wal_level=minimal is set.
-- Because on mirrored cluster primary segments have replication slots
-- and that conflict with the wal_level=minimal GUC
-- we connect to coordinator in utility mode for AO/AOCO operations and
-- validate WAL records on the coordinator.

GP_IGNORE: formatted by atmsort.pm
-- start_matchignore
-- m/pg_xlogdump: FATAL:  error in WAL record at */
-- m/.*The 'DISTRIBUTED BY' clause determines the distribution of data*/
-- m/.*Table doesn't have 'DISTRIBUTED BY' clause*/
-- end_matchignore
GP_IGNORE: defined new match expression

-- Create tables (AO, AOCO)
-1U: CREATE TABLE ao_foo (n int) WITH (appendonly=true);
CREATE
-1U: CREATE TABLE aoco_foo (n int, m int) WITH (appendonly=true, orientation=column);
CREATE

-- Switch WAL file
-1U: SELECT true FROM pg_switch_xlog();
 bool 
------
 t    
(1 row)
-- Insert data (AO)
-1U: INSERT INTO ao_foo SELECT generate_series(1,10);
INSERT 10
-- Insert data (AOCO)
-1U: INSERT INTO aoco_foo SELECT generate_series(1,10), generate_series(1,10);
INSERT 10
-- Delete data and run vacuum (AO)
-1U: DELETE FROM ao_foo WHERE n > 5;
DELETE 5
-1U: VACUUM ao_foo;
VACUUM
-- Delete data and run vacuum (AOCO)
-1U: DELETE FROM aoco_foo WHERE n > 5;
DELETE 5
-1U: VACUUM aoco_foo;
VACUUM
-1Uq: ... <quitting>

-- Validate wal records (mirrorless setting has alternative answer file for this since wal_level is already minimal)
! last_wal_file=$(psql -At -c "SELECT pg_xlogfile_name(pg_current_xlog_location())" postgres) && pg_xlogdump ${last_wal_file} -p ${MASTER_DATA_DIRECTORY}/pg_xlog -r appendonly;
rmgr: Appendonly  len (rec/tot):    160/   192, tx:        741, lsn: 0/10000138, prev 0/100000E8, bkp: 0000, desc: insert: rel 1663/16396/16387 seg/offset:0/0 len:136
rmgr: Appendonly  len (rec/tot):     24/    56, tx:        742, lsn: 0/10000380, prev 0/100002F0, bkp: 0000, desc: insert: rel 1663/16396/16391 seg/offset:128/0 len:0
rmgr: Appendonly  len (rec/tot):    104/   136, tx:        742, lsn: 0/10000408, prev 0/100003B8, bkp: 0000, desc: insert: rel 1663/16396/16391 seg/offset:0/0 len:80
rmgr: Appendonly  len (rec/tot):    104/   136, tx:        742, lsn: 0/10000490, prev 0/10000408, bkp: 0000, desc: insert: rel 1663/16396/16391 seg/offset:128/0 len:80
rmgr: Appendonly  len (rec/tot):     24/    56, tx:          0, lsn: 0/100006F8, prev 0/100006C8, bkp: 0000, desc: truncate: rel 1663/16396/16387 seg/offset:0/136
rmgr: Appendonly  len (rec/tot):     24/    56, tx:          0, lsn: 0/10040DF0, prev 0/10040DC0, bkp: 0000, desc: truncate: rel 1663/16396/16391 seg/offset:0/80
rmgr: Appendonly  len (rec/tot):     24/    56, tx:          0, lsn: 0/10040E28, prev 0/10040DF0, bkp: 0000, desc: truncate: rel 1663/16396/16391 seg/offset:128/80
pg_xlogdump: FATAL:  error in WAL record at 0/100813D8: record with zero length at 0/10081408


-- *********** Set wal_level=minimal **************
!\retcode gpconfig -c wal_level -v minimal --masteronly --skipvalidation;
(exited with code 0)
-- Set max_wal_senders to 0 because a non-zero value requires wal_level >= 'archive'
!\retcode gpconfig -c max_wal_senders -v 0 --masteronly --skipvalidation;
(exited with code 0)
-- Restart QD
!\retcode pg_ctl -l /dev/null -D $MASTER_DATA_DIRECTORY restart -w -t 600 -m fast;
(exited with code 0)

-- Switch WAL file
-1U: SELECT true FROM pg_switch_xlog();
 bool 
------
 t    
(1 row)
-- Insert data (AO)
-1U: INSERT INTO ao_foo SELECT generate_series(1,10);
INSERT 10
-- Insert data (AOCO)
-1U: INSERT INTO aoco_foo SELECT generate_series(1,10), generate_series(1,10);
INSERT 10
-- Delete data and run vacuum (AO)
-1U: DELETE FROM ao_foo WHERE n > 5;
DELETE 5
-1U: VACUUM ao_foo;
VACUUM
-- Delete data and run vacuum (AOCO)
-1U: DELETE FROM aoco_foo WHERE n > 5;
DELETE 5
-1U: VACUUM aoco_foo;
VACUUM

-- Validate wal records
! last_wal_file=$(psql -At -c "SELECT pg_xlogfile_name(pg_current_xlog_location())" postgres) && pg_xlogdump ${last_wal_file} -p ${MASTER_DATA_DIRECTORY}/pg_xlog -r appendonly;


-1U: DROP TABLE ao_foo;
DROP
-1U: DROP TABLE aoco_foo;
DROP

-- Reset wal_level
!\retcode gpconfig -r wal_level --masteronly --skipvalidation;
(exited with code 0)
-- Reset max_wal_senders
!\retcode gpconfig -r max_wal_senders --masteronly --skipvalidation;
(exited with code 0)
-- Restart QD
!\retcode pg_ctl -l /dev/null -D $MASTER_DATA_DIRECTORY restart -w -t 600 -m fast;
(exited with code 0)
